README.markdown
# copyEvents

A jQuery plugin that provides methods for copying events (including the namespaces and data) from one element to another.

There are two methods, `copyEvents` and `copyEventTo`. They both take a selector, jQuery collection, or just an element reference. 

    // Copies events to the main selection from the elements passed to .copyEvents
    $('#copy_events_to_me').copyEvents('#from_me');
    
    // Copies events from the main selection to the elements passed to .copyEventsTo
    $('#copy_events_from_me').copyEventsTo('#to_me');

This plugin is tested against 1.0.4, 1.1.4, 1.2.6, 1.3.2, 1.4.4, 1.5.2, 1.6.4, 1.7.2, 1.8.2, and 1.9.1

## License

The copyEvents plugin is licensed under the MIT License (LICENSE.txt).

Copyright (c) 2013 [Brandon Aaron](http://brandonaaron.net)


ChangeLog.markdown
# New and Noteworthy

# 1.4

* Made compatiable with AMD loaders
* Updated test suite
* jQuery plugins manifest

# 1.3

* Fixed compatibility with jQuery 1.8

# 1.2

* Fixed compatibility with jQuery 1.4.2
* Fixed compatibility with older version of jQuery as well (Only tested against the latest version in each branch older than 1.4)

# 1.1

* Removed cloneWithEvents in favor of the new clone(true) method in the core of 1.2


# 1.0.1

* Updated cloneWithEvents to clone events on descendants when doing a deep clone


# 1.0

* Updated to work with jQuery 1.1.2+
* Added cloneWithEvents to clone elements and copy it's events at the same time


test/test.js
function testWithVersion(version/*args*/) {
  var args = Array.prototype.slice.call(arguments, 1);
  args[0] = args[0] + '(' + version + '+)';
  if (isVersionGreater($.fn.jquery, version)) {
    test.apply(window, args);
  }

  function isVersionGreater(v1, v2) {
    var v1 = parseInt(v1.split('.').join(''), 10),
        v2 = parseInt(v2.split('.').join(''), 10);
    return (v1 >= v2);
  }
}

// should work with all versions of jQuery
testWithVersion('1.0.4', "copyEvents()", function() {
    var clickCount = mouseoverCount = mouseoutCount = 0;

    // Create event handlers for A tag
    $('#main a.from')
        .bind('click', function() {
            clickCount++;
        })
        .bind('mouseover', function() {
            mouseoverCount++;
        })
        .bind('mouseout', function() {
            mouseoutCount++;
        });

    // clone events from a to 'from' li tags
    $('#main li.from')
        .copyEvents('#main a.from')
        // fire events
        .trigger('click')
        .trigger('mouseover')
        .trigger('mouseout');

    // check counts
    ok( clickCount == 2, 'Testing click count' );
    ok( mouseoverCount == 2, 'Testing mouseover count' );
    ok( mouseoutCount == 2, 'Testing mouseout count' );

    // just making sure there are no errors when copying to or from an empty set or from elements with no events
    $('#does_not_exist').copyEvents('#also_does_not_exist');
    $('<div />').copyEvents('#does_not_exist');
});

// should work with all versions of jQuery
testWithVersion('1.0.4', "copyEventsTo()", function() {
    var clickCount = mouseoverCount = mouseoutCount = 0;

    // Create event handlers for A tag
    $('#main a.to')
        .bind('click', function() {
            clickCount++;
        })
        .bind('mouseover', function() {
            mouseoverCount++;
        })
        .bind('mouseout', function() {
            mouseoutCount++;
        });

    // clone events from a to 'from' li tags
    $('#main a.to')
        .copyEventsTo('#main li.to');

    $('#main li.to')
        // fire events
        .trigger('click')
        .trigger('mouseover')
        .trigger('mouseout');

    // check counts
    ok( clickCount == 2, 'Testing click count' );
    ok( mouseoverCount == 2, 'Testing mouseover count' );
    ok( mouseoutCount == 2, 'Testing mouseout count' );
});

// should work since 1.1.x when data was introduced
testWithVersion('1.1.4', "copyEvents() with data", function() {
    var clickCount = mouseoverCount = mouseoutCount = dataCount = 0;

    // Create event handlers for A tag
    $('#main a.from')
        .bind('click', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            clickCount++;
        })
        .bind('mouseover', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoverCount++;
        })
        .bind('mouseout', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoutCount++;
        });

    // clone events from a to 'from' li tags
    $('#main li.from')
        .copyEvents('#main a.from')
        // fire events
        .trigger('click')
        .trigger('mouseover')
        .trigger('mouseout');

    // check counts
    ok( clickCount == 2, 'Testing click count' );
    ok( mouseoverCount == 2, 'Testing mouseover count' );
    ok( mouseoutCount == 2, 'Testing mouseout count' );
    ok( dataCount == 6, 'Testing data was copied' );
});

// should work since 1.1.x when data was introduced
testWithVersion('1.1.4', "copyEvents() with data", function() {
    var clickCount = mouseoverCount = mouseoutCount = dataCount = 0;

    // Create event handlers for A tag
    $('#main a.to')
        .bind('click', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            clickCount++;
        })
        .bind('mouseover', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoverCount++;
        })
        .bind('mouseout', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoutCount++;
        });

    // clone events from a to 'from' li tags
    $('#main a.to')
        .copyEventsTo('#main li.to');

    $('#main li.to')
        // fire events
        .trigger('click')
        .trigger('mouseover')
        .trigger('mouseout');

    // check counts
    ok( clickCount == 2, 'Testing click count' );
    ok( mouseoverCount == 2, 'Testing mouseover count' );
    ok( mouseoutCount == 2, 'Testing mouseout count' );
    ok( dataCount == 6, 'Testing data was copied' );
});

// should work since 1.2.x when namespaced events were introduced
testWithVersion('1.2.6', "copyEvents() with namespaces", function() {
    var clickCount = mouseoverCount = mouseoutCount = dataCount = 0;

    // Create event handlers for A tag
    $('#main a.from')
        .bind('click.yay.query', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            clickCount++;
        })
        .bind('mouseover.yay.query', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoverCount++;
        })
        .bind('mouseout.yay.query', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoutCount++;
        });

    // clone events from a to 'from' li tags
    $('#main li.from')
        .copyEvents('#main a.from')
        // fire events
        .trigger('click.yay.query')
        .trigger('mouseover.yay.query')
        .trigger('mouseout.yay.query');

    // check counts
    ok( clickCount == 2, 'Testing click count' );
    ok( mouseoverCount == 2, 'Testing mouseover count' );
    ok( mouseoutCount == 2, 'Testing mouseout count' );
    ok( dataCount == 6, 'Testing data was copied' );
});

// should work since 1.2.x when namespaced events were introduced
testWithVersion('1.2.6', "copyEvents() with namespaces", function() {
    var clickCount = mouseoverCount = mouseoutCount = dataCount = 0;

    // Create event handlers for A tag
    $('#main a.to')
        .bind('click.yay.query', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            clickCount++;
        })
        .bind('mouseover.yay.query', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoverCount++;
        })
        .bind('mouseout.yay.query', { test: true }, function(event) {
            if (event.data.test) dataCount++;
            mouseoutCount++;
        });

    // clone events from a to 'from' li tags
    $('#main a.to')
        .copyEventsTo('#main li.to');

    $('#main li.to')
        // fire events
        .trigger('click.yay.query')
        .trigger('mouseover.yay.query')
        .trigger('mouseout.yay.query');

    // check counts
    ok( clickCount == 2, 'Testing click count' );
    ok( mouseoverCount == 2, 'Testing mouseover count' );
    ok( mouseoutCount == 2, 'Testing mouseout count' );
    ok( dataCount == 6, 'Testing data was copied' );
});


jquery.copyEvents.js
/*! Copyright (c) 2013 Brandon Aaron (http://brandonaaron.net)
 * Licensed under the MIT License (LICENSE.txt).
 *
 * Version: 1.4
 */
(function (factory) {
    if (typeof define === 'function' && define.amd) {
        // AMD. Register as an anonymous module.
        define(['jquery'], factory);
    } else {
        // Browser globals
        factory(jQuery);
    }
}(function ($) {
   $.fn.extend({
        copyEvents: function(from) {
            $.event.copy($(from), this);
            return this;
        },

        copyEventsTo: function(to) {
            $.event.copy(this, $(to));
            return this;
        }
    });

    $.event.copy = function(from, to) {
        // $._data: 1.8 branch
        // $.data wasn't introduced until the 1.2 branch
        // 1.1 branch stored event data on the element using "$events" expando
        // 1.0 branch stored event data on the element using "events" expando
        var fromEl = from[0],
            events = fromEl && (($._data && $._data(fromEl, 'events')) || ($.data && $.data(fromEl, 'events')) || fromEl.$events || fromEl.events) || {};

        to.each(function() {
            var elem = this;
            for (var type in events) {
                // in jQuery 1.4.2+ event handlers are stored in an array, previous versions it is an object
                $.each(events[type], function(index, handler) {
                    // in jQuery 1.4.2+ handler is an object with a handle property
                    // in previous versions it is the actual handler with the properties tacked on
                    var namespaces = handler.namespace !== undefined && handler.namespace || handler.type || '';
                    namespaces = namespaces.length ? (namespaces.indexOf('.') === 0 ? '' : '.') + namespaces : '';
                    $.event.add(elem, type + namespaces, handler.handler || handler, handler.data);
                });
            }
        });
    };
}));
